scico/solver.py

```diff
@@ -229,25 +230,36 @@ def minimize(
         min_func = _wrap_func(func_, x0_shape, x0_dtype)
         jac = False
 
-    res = spopt.minimize(
-        min_func,
-        x0=x0,
-        args=args,
-        jac=jac,
-        method=method,
-        options=options,
+    res_dict = {}
+
+    def fun(_):
+        min_res = spopt.minimize(
+            min_func,
+            x0=x0,
+            args=args,
+            jac=jac,
+            method=method,
+            options=options,
+        )  # Returns OptimizeResult
+        res_dict.update({"result": min_res})  # Updates res dict with side effect
+        return min_res.x.astype(x0.dtype)  # Return for host_callback
+
+    # hcb call with side effects to get the OptimizeResult on the same device it was called
+    hcb.call(
+        fun,
+        arg=None,
+        result_shape=x0,
+        call_with_device=isinstance(x0, jax.interpreters.xla.DeviceArray),
     )
 
+    res = res_dict["result"]
     # un-vectorize the output array, put on device
     res.x = snp.reshape(
         res.x, x0_shape
     )  # if x0 was originally a BlockArray then res.x is converted back to one here
 
     res.x = res.x.astype(x0_dtype)
```

Comment:
Duplicates line 245

Reply:
This line is so that I could extract the side effect created by fun out of the dictionary for readability (using res instead of res_dict["result"] in the later parts of the method. Should I still remove it and fix it to use that?

Reply:
Addressed in the other comment, cleaned up syntax.